neqo-transport/src/crypto.rs

```diff
@@ -916,11 +921,60 @@ impl CryptoStates {
         }
     }
 
+    /// When decrypting Initial packets, there are potentially multiple active versions.
+    /// The `used_pn` range tracks what has been received on the version that was used.
+    /// But if the version changes, the version we select might have a value of 0,
+    /// rather than the actual value, which can cause packet number recovery to fail.
+    /// To avoid that, have the indicated `version` continue from the previous version.
+    /// This only needs to be run once, so run it when getting header protection.
+    fn maybe_continue_initial_rx(&mut self, version: Version) {
+        // Only do this if this version hasn't been used...
+        // This would be better with `is_none_or`, but that needs rust 1.82
+        if !self.initials[version]
+            .as_ref()
+            .is_some_and(|dx| dx.rx.next_pn() == 0)
+        {
+            return;
+        }
+        // ... and some other version has been.
+        // This assumes that there is just one other version in use,
+        // as the spec requires.
+        let Some(other) = self
+            .initials
+            .iter()
+            .find_map(|(k, v)| v.as_ref().is_some_and(|z| z.rx.next_pn() > 0).then_some(k))
+        else {
+            return;
+        };
+        debug_assert_ne!(version, other);
+
+        // This uses the take-modify-restore pattern to avoid
+        // having the borrow checker complain.
+        // It *ignores* errors from the `continuation()`
+        // so that the restore step isn't skipped.
+        //
+        // This doesn't need to be full anti-replay.
+        // Each version has separate keys, so nonce reuse is OK.
+        // After this, we might reject packets if the peer
+        // does reuse nonces, but they aren't allowed to do that.
+        //
+        // Note: these `if let Some(...)` conditions are always true.
```

Comment:
If so, why not `expect()` instead? Would allow us to catch if they ever aren't true.

Reply:
This is literally impossible: see above where each is checked with `is_some_and()`.  The reason I didn't use expect is that it is more code in this case.  `self.initials[whatever].as_ref().expect("whatever").rx` or `self.initials[whatever].take().expect("whatever").rx` is nearly as much code all told, with the added ugliness of expect.